Returning Self Considered Harmful

randomPoison opened the pull request.

randomPoison committed:
Create returning-self.md

randomPoison committed:
Test out adding code snippets

randomPoison committed:
Add more stufffffffff

randomPoison committed:
Update complete demo link

randomPoison committed:
Split method chaining into its own article

randomPoison committed:
Cleanup single chain example

randomPoison committed:
Cleanup intermediate example

randomPoison committed:
Fixup example

randomPoison committed:
Change header style

randomPoison committed:
Add summary to first example section

randomPoison committed:
Add summary for use before consuming

randomPoison committed:
Update returning-self.md

randomPoison committed:
Start writing conclusion

randomPoison committed:
Update returning-self.md

randomPoison committed:
Update returning-self.md

randomPoison committed:
Spell check pass

randomPoison committed:
Add section on cascading

randomPoison committed:
Remove method-chaining draft

randomPoison committed:
Editing pass and cleanup

Velfi reviewed (changes requested):
typo 'By binding the intial'

randomPoison committed:
Fix typo

randomPoison committed:
Fix typos and such

randomPoison committed:
Add playground links for code examples

randomPoison committed:
Change name to "Returning Self Considered Harmful"

randomPoison commented:
Lokathor on Discord brought up the point that if you do a more functional style, where all data is immutable, it's natural for all methods to take `self` and return `Self`, returning a copy of the input data with the modification applied.

It might be worth addressing this case in the post. And example of this is iterator adapters: They take `self` and return the new iterator by value, and therefore work smoothly as a method chain. The think to clarify in the post is that it's totally valid to return `Self` *as long as it's not solely to enable method chaining*. As long as returning `Self` is actually a meaningful part of your API, it's entirely valid to do so.

Velfi commented:
Just be friendly and remember that you're trying to help people achieve their goals.

randomPoison committed:
First pass at clarifying language

* Change name to "Chaining Functions Without Returning Self".
* Add summary of key points to opening section.
* Reorganize the various examples under a subheading.
* Add a section to elaborate on appropriate cases for returning `Self`.
* Rename "It's Ultimately a Hack" section match key point.

randomPoison committed:
Reword post to focus on returning self

Switch the focus from returning any variant of `Self` to specifically returning the `self` parameter at the end of a function. This clears up some concerns that reviewers had voiced, and more directly addresses the problematic cases (since there are lots of valid reasons to return `Self` but far fewer times when returning `self` makes sense).

randomPoison committed:
Add a disclaimer section

randomPoison committed:
Cleanup the conclusion and such

randomPoison committed:
Setup permalink for post

randomPoison committed:
Merge branch 'master' into draft/returning-self

randomPoison committed:
Final pass at rewording and clarification
